<a href="https://colab.research.google.com/github/devrave/Analisis_salarios_data_2025/blob/main/analitica_salarios.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Determinantes del salario en profesionales de datos (2025)

Este notebook analiza los factores que explican el salario de los profesionales de datos usando la encuesta Stack Overflow Developer Survey 2025.

## 1. Importar librerías necesarias

En este paso importamos las librerías que usaremos para análisis de datos y visualización. Explicaremos para qué sirve cada una.

In [4]:
# pandas: para manipulación y análisis de datos
# numpy: para operaciones numéricas
# matplotlib y seaborn: para visualización de datos
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from IPython.display import display, Markdown

# Configuración de gráficos para que se vean bien en el notebook
# La función sns.set() está deprecada, así que usamos set_theme()
sns.set_theme(style="whitegrid")
plt.rcParams['figure.figsize'] = (10, 6)

print("Librerías importadas correctamente ✔️")

Librerías importadas correctamente ✔️


In [5]:
# Configuración global para mostrar todos los datos en pandas
import pandas as pd
pd.set_option('display.max_rows', None)  # Muestra todas las filas
pd.set_option('display.max_columns', None)  # Muestra todas las columnas
pd.set_option('display.width', None)  # No limita el ancho de la consola
pd.set_option('display.max_colwidth', None)  # Muestra el contenido completo de cada celda

## 2. Cargar y explorar los datos

En este paso vamos a cargar el archivo survey_results_public.csv y a explorar las primeras filas para entender la estructura de los datos. Explicaremos cada línea de código.

In [6]:
# Cargar el archivo CSV en un DataFrame de pandas
# El parámetro encoding='utf-8' es estándar, pero si da error prueba con 'latin1' o 'ISO-8859-1'
df = pd.read_csv('/content/drive/MyDrive/Analisis_de_datos/Proyecto_analitica/survey_results_public.csv', encoding='utf-8')

print("Datos cargados correctamente ✔️")

# Mostrar las primeras 5 filas del DataFrame para ver la estructura de los datos
df.head()

/tmp/ipython-input-3259010491.py:3: DtypeWarning: Columns (56,74,92,97,98,105,109,110,132,162,165) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv('/content/drive/MyDrive/Analisis_de_datos/Proyecto_analitica/survey_results_public.csv', encoding='utf-8')


Datos cargados correctamente ✔️


ResponseId                      MainBranch              Age  \
0           1  I am a developer by profession  25-34 years old   
1           2  I am a developer by profession  25-34 years old   
2           3  I am a developer by profession  35-44 years old   
3           4  I am a developer by profession  35-44 years old   
4           5  I am a developer by profession  35-44 years old   

                                           EdLevel  \
0  Master’s degree (M.A., M.S., M.Eng., MBA, etc.)   
1              Associate degree (A.A., A.S., etc.)   
2     Bachelor’s degree (B.A., B.S., B.Eng., etc.)   
3     Bachelor’s degree (B.A., B.S., B.Eng., etc.)   
4  Master’s degree (M.A., M.S., M.Eng., MBA, etc.)   

                                             Employment  \
0                                              Employed   
1                                              Employed   
2  Independent contractor, freelancer, or self-employed   
3                                              Employed   
4  Independent contractor, freelancer, or self-employed   

                                    EmploymentAddl  WorkExp  \
0  Caring for dependents (children, elderly, etc.)      8.0   
1                                              NaN      2.0   
2                                None of the above     10.0   
3                                None of the above      4.0   
4  Caring for dependents (children, elderly, etc.)     21.0   

                                                                               LearnCodeChoose  \
0    Yes, I am not new to coding but am learning new coding techniques or programming language   
1    Yes, I am not new to coding but am learning new coding techniques or programming language   
2    Yes, I am not new to coding but am learning new coding techniques or programming language   
3    Yes, I am not new to coding but am learning new coding techniques or programming language   
4  No, I am not new to coding and did not learn new coding techniques or programming languages   

                                                                                                                                                                                                                                                                           LearnCode  \
0                                                                                                                                                  Online Courses or Certification (includes all media types);Other online resources (e.g. standard search, forum, online community)   
1                     Online Courses or Certification (includes all media types);Other online resources (e.g. standard search, forum, online community);Books / Physical media;Videos (not associated with specific online course or certification);Stack Overflow or Stack Exchange   
2                                                                                   Online Courses or Certification (includes all media types);Videos (not associated with specific online course or certification);Technical documentation (is generated for/by the tool or system)   
3  Other online resources (e.g. standard search, forum, online community);Videos (not associated with specific online course or certification);Stack Overflow or Stack Exchange;Technical documentation (is generated for/by the tool or system);AI CodeGen tools or AI-enabled apps   
4                                                                                                                                                                                                                                                                                NaN   

                                                                           LearnCodeAI  \
0  Yes, I learned how to use AI-enabled tools for my personal curiosity and/or hobbies   
1  Yes, I learned how to use AI-enabled tools for my personal curiosity and/or hobbies   
2  Yes,

In [2]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


### 2.1. Exploración inicial de las columnas y tipos de datos

Ahora que cargamos los datos, vamos a ver qué columnas tiene el archivo, qué tipo de información contiene cada una y cómo están representadas las variables clave para nuestro análisis.

Esto nos ayudará a identificar si necesitamos limpiar, transformar o seleccionar variables antes de cualquier análisis profundo.

#### ¿Qué columnas tiene el DataFrame?
Esto nos permite ver todas las variables disponibles para el análisis y su orden.

In [ ]:
# Mostramos el listado de columnas del DataFrame
display(df.columns.tolist())

['ResponseId',
 'MainBranch',
 'Age',
 'EdLevel',
 'Employment',
 'EmploymentAddl',
 'WorkExp',
 'LearnCodeChoose',
 'LearnCode',
 'LearnCodeAI',
 'AILearnHow',
 'YearsCode',
 'DevType',
 'OrgSize',
 'ICorPM',
 'RemoteWork',
 'PurchaseInfluence',
 'TechEndorseIntro',
 'TechEndorse_1',
 'TechEndorse_2',
 'TechEndorse_3',
 'TechEndorse_4',
 'TechEndorse_5',
 'TechEndorse_6',
 'TechEndorse_7',
 'TechEndorse_8',
 'TechEndorse_9',
 'TechEndorse_13',
 'TechEndorse_13_TEXT',
 'TechOppose_1',
 'TechOppose_2',
 'TechOppose_3',
 'TechOppose_5',
 'TechOppose_7',
 'TechOppose_9',
 'TechOppose_11',
 'TechOppose_13',
 'TechOppose_16',
 'TechOppose_15',
 'TechOppose_15_TEXT',
 'Industry',
 'JobSatPoints_1',
 'JobSatPoints_2',
 'JobSatPoints_3',
 'JobSatPoints_4',
 'JobSatPoints_5',
 'JobSatPoints_6',
 'JobSatPoints_7',
 'JobSatPoints_8',
 'JobSatPoints_9',
 'JobSatPoints_10',
 'JobSatPoints_11',
 'JobSatPoints_13',
 'JobSatPoints_14',
 'JobSatPoints_15',
 'JobSatPoints_16',
 'JobSatPoints_15_TEXT',
 

#### ¿Qué tipo de datos tiene cada columna?
Esto ayuda a identificar si alguna variable requiere conversión de tipo o tratamiento especial. También haremos un loop para identificar el numero de la columna e identificarla mas rápido.


In [7]:
# Mostramos los tipos de datos de cada columna y le vamos agregar una numeracion para facilitar la referencia
print("Tipos de datos de cada columna:")
print("-" * 40)
for i, col in enumerate(df.columns,start=1):
    print(f"{i}: {col} - {df[col].dtype}")

Tipos de datos de cada columna:
----------------------------------------
1: ResponseId - int64
2: MainBranch - object
3: Age - object
4: EdLevel - object
5: Employment - object
6: EmploymentAddl - object
7: WorkExp - float64
8: LearnCodeChoose - object
9: LearnCode - object
10: LearnCodeAI - object
11: AILearnHow - object
12: YearsCode - float64
13: DevType - object
14: OrgSize - object
15: ICorPM - object
16: RemoteWork - object
17: PurchaseInfluence - object
18: TechEndorseIntro - object
19: TechEndorse_1 - float64
20: TechEndorse_2 - float64
21: TechEndorse_3 - float64
22: TechEndorse_4 - float64
23: TechEndorse_5 - float64
24: TechEndorse_6 - float64
25: TechEndorse_7 - float64
26: TechEndorse_8 - float64
27: TechEndorse_9 - float64
28: TechEndorse_13 - float64
29: TechEndorse_13_TEXT - object
30: TechOppose_1 - float64
31: TechOppose_2 - float64
32: TechOppose_3 - float64
33: TechOppose_5 - float64
34: TechOppose_7 - float64
35: TechOppose_9 - float64
36: TechOppose_11 - float64
3

#### Resumen estadístico de las variables numéricas
Permite ver rápidamente la media, desviación estándar, valores mínimos y máximos, y detectar posibles outliers.

In [ ]:
# Mostramos un resumen estadístico de las variables numéricas
display(df.describe(include='all'))

ResponseId                      MainBranch              Age  \
count   49191.000000                           49191            49191   
unique           NaN                               6                7   
top              NaN  I am a developer by profession  25-34 years old   
freq             NaN                           37467            16519   
mean    24596.000000                             NaN              NaN   
std     14200.362883                             NaN              NaN   
min         1.000000                             NaN              NaN   
25%     12298.500000                             NaN              NaN   
50%     24596.000000                             NaN              NaN   
75%     36893.500000                             NaN              NaN   
max     49191.000000                             NaN              NaN   

                                             EdLevel Employment  \
count                                          48149      48339   
unique                                             8          6   
top     Bachelor’s degree (B.A., B.S., B.Eng., etc.)   Employed   
freq                                           20278      33750   
mean                                             NaN        NaN   
std                                              NaN        NaN   
min                                              NaN        NaN   
25%                                              NaN        NaN   
50%                                              NaN        NaN   
75%                                              NaN        NaN   
max                                              NaN        NaN   

           EmploymentAddl       WorkExp  \
count               44875  42893.000000   
unique                171           NaN   
top     None of the above           NaN   
freq                18613           NaN   
mean                  NaN     13.367403   
std                   NaN     10.800117   
min                   NaN      1.000000   
25%                   NaN      5.000000   
50%                   NaN     10.000000   
75%                   NaN     20.000000   
max                   NaN    100.000000   

                                                                                  LearnCodeChoose  \
count                                                                                       46858   
unique                                                                                          3   
top     Yes, I am not new to coding but am learning new coding techniques or programming language   
freq                                                                                        32362   
mean                                                                                          NaN   
std                                                                                           NaN   
min                                                                                           NaN   
25%                                                                                           NaN   
50%                                                                                           NaN   
75%                                                                                           NaN   
max                                                                                           NaN   

                                                                                                                                                                       LearnCode  \
count                                                                                                                                                                      33556   
unique                                                                                                                                                                      2767   
top     Other online resources (e.g. standard search, foru

#### Valores nulos por columna
Revisar los valores nulos ayuda a identificar posibles problemas de calidad de datos y a planear estrategias de limpieza o imputación.

In [ ]:
# Valores nulos por columna
display(df.isnull().sum())

ResponseId                                      0
MainBranch                                      0
Age                                             0
EdLevel                                      1042
Employment                                    852
EmploymentAddl                               4316
WorkExp                                      6298
LearnCodeChoose                              2333
LearnCode                                   15635
LearnCodeAI                                  3990
AILearnHow                                  20934
YearsCode                                    6149
DevType                                      5511
OrgSize                                     15013
ICorPM                                      15948
RemoteWork                                  15411
PurchaseInfluence                           11754
TechEndorseIntro                            11635
TechEndorse_1                               13216
TechEndorse_2                               13216


In [ ]:
# Porcentaje de valores nulos en las principales columnas
print("Porcentaje de valores nulos en las principales columnas:")
columnas_principales = ['Country', 'RemoteWork', 'OrgSize', 'Employment', 'DevType', 'Industry']
for col in columnas_principales:
    if col in df.columns:
        porcentaje_nulos = df[col].isnull().mean() * 100
        print(f"  {col}: {porcentaje_nulos:.2f}% nulos")
    else:
        print(f"  {col}: columna no encontrada en el dataset")

Porcentaje de valores nulos en las principales columnas:
  Country: 27.96% nulos
  RemoteWork: 31.33% nulos
  OrgSize: 30.52% nulos
  Employment: 1.73% nulos
  DevType: 11.20% nulos
  Industry: 31.61% nulos


Conclusiones basadas en el análisis de valores nulos

### Resultados que encontramos:
- **Country**: 27.96% nulos
- **RemoteWork**: 31.33% nulos
- **OrgSize**: 30.52% nulos
- **Employment**: 1.73% nulos
- **DevType**: 11.20% nulos
- **Industry**: 31.61% nulos

---

### Mi análisis y propuesta para el equipo con respecto a los valores nulos:

#### ✅ **Aquí no hay problema, sigamos adelante sin dudas:**

**Employment (1.73% nulos)**: Chicos, esta variable está prácticamente completa. Solo el 1.73% falta, así que creó que podemos usarla sin preocupaciones. Si queremos ser muy rigurosos, podemos simplemente eliminar esas pocas filas con datos faltantes o rellenarlas con la categoría más frecuente.

**DevType (11.20% nulos)**: Esta tampoco es un problema importante. Con apenas 11% de datos faltantes, propongo que impute estos valores con la categoría más frecuente (moda). El impacto en el análisis será mínimo y ganamos más datos para trabajar.

#### ⚠️ **Aquí tenemos que tomar una decisión como equipo:**

**Country (27.96% nulos)**, **RemoteWork (31.33% nulos)**, **OrgSize (30.52% nulos)** e **Industry (31.61% nulos)**: Chicos, estas cuatro variables tienen entre 28% y 31% de datos faltantes. Eso es bastante. Entonces, para cada una de ellas, yo propongo que nos hagamos la pregunta: **¿Es esta variable realmente importante para explicar los determinantes del salario?**

Si la respuesta es **SÍ**, entonces yo propongo que mantengamos la columna e impute los valores nulos con la categoría más frecuente (la moda). Así no perdemos información.

Si la respuesta es **NO**, entonces creo que deberíamos eliminar esa columna para mejorar la calidad general del análisis.

Por ejemplo:
- Si creemos que el **País** es fundamental para entender por qué varía el salario, entonces mantengámosla e impute.
- Si creemos que la **Modalidad de trabajo (remoto/presencial)** es clave, mantengámosla e impute.
- Si la **Industria** es importante para nuestro análisis, manténgamosla e impute.
- Si el **Tamaño de la empresa** (OrgSize) es relevante, manténgamosla e impute.

### Mi recomendación final para la reunión:

1. **Definitivamente usaremos**: Employment y DevType. No debemos perder esas variables.
2. **Tenemos que discutir**: Country, RemoteWork, OrgSize e Industry. Debemos definir juntos cuáles son realmente importantes para nuestro objetivo de investigación.
3. **Acción inmediata**: Para cualquier columna que decidamos mantener de ese grupo, yo propongo que impute los valores faltantes con la moda (la categoría más frecuente). Es la forma más simple y razonable.
4. **Documentación**: Dejaremos escrito en nuestro reporte exactamente qué decision tomamos para cada columna y por qué. Eso nos ahorra explicaciones después.

#### ¿Cuántos valores únicos tiene cada columna?
Este análisis permite identificar variables con baja variabilidad (posibles candidatas a eliminar) o con demasiadas categorías, lo que puede complicar el análisis o la modelización.

In [ ]:
# Cantidad de valores únicos por columna
display(df.nunique(dropna=False))

ResponseId                                  49191
MainBranch                                      6
Age                                             7
EdLevel                                         9
Employment                                      7
EmploymentAddl                                172
WorkExp                                        73
LearnCodeChoose                                 4
LearnCode                                    2768
LearnCodeAI                                     6
AILearnHow                                   1678
YearsCode                                      79
DevType                                        33
OrgSize                                        10
ICorPM                                          3
RemoteWork                                      6
PurchaseInfluence                               6
TechEndorseIntro                                4
TechEndorse_1                                  15
TechEndorse_2                                  15


### 2.2. Análisis exploratorio inicial
En este paso realizaremos un análisis exploratorio general para conocer mejor los datos. Nos enfocaremos en:
- Identificar valores únicos y distribución en variables clave (país, modalidad de trabajo, tamaño de empresa, rol, industria).
- Detectar datos faltantes y posibles valores atípicos.
- Explorar la variable objetivo (salario) y su relación con otras variables.

**Nota:** Ahora, gracias a la configuración global de pandas, todos los resultados de conteos y tablas se mostrarán completos, sin recortes. Esto facilita la revisión exhaustiva de los datos.

Este análisis nos permitirá entender la estructura, calidad y potencial de los datos antes de avanzar a modelos o visualizaciones más complejas.

In [ ]:
# Análisis de valores únicos y distribución en variables clave
# Reemplaza los nombres de columna según corresponda en tu dataset
variables_clave = [
    'Country', # País
    'RemoteWork', # Modalidad de trabajo
    'OrgSize', # Tamaño de empresa
    'Employment', # Tipo de empleo
    'DevType', # Rol actual
    'Industry', # Industria
]

for var in variables_clave:
    if var in df.columns:
        print(f"\n--- {var}: valores únicos y distribución ---")
        print(df[var].value_counts(dropna=False).to_string())
    else:
        print(f"\n--- {var}: columna no encontrada en el dataset ---")


--- Country: valores únicos y distribución ---
Country
NaN                                                     13754
United States of America                                 7233
Germany                                                  3025
India                                                    2547
United Kingdom of Great Britain and Northern Ireland     2042
France                                                   1409
Canada                                                   1305
Ukraine                                                   964
Poland                                                    888
Netherlands                                               867
Italy                                                     835
Brazil                                                    825
Australia                                                 804
Spain                                                     717
Sweden                                                    616
Switzerland   